File: pruefstuecke/profil.mjs
/**
 * Prüfstück: Das Profil eines Menschen – das eigene und das der Kollegen.
 *
 * DER AUFTRAG (Inhaber, 09.09.2026): „jetzt alles was in Portal
 * eingetragen wird pro User soll in der App im Profil angezeogt werden,
 * soll im einegnen Profil als auch im Profil andere Koolegen der
 * Organistation." Und der Satz davor, der sagt WOFÜR: „vorallem um in
 * Apps nach der mitarbeiter zu suchen und anzuchaten wie bei Whatsapp
 * oder Teams."
 *
 * DER SCHADEN: Die Kollegenliste kannte vier Felder – id, name, nummer,
 * durchwahl. Wer jemanden suchte, musste seinen NAMEN kennen. In einem
 * Haus mit vierzig Leuten findet man so den Kollegen aus der
 * Buchhaltung nicht, dessen Namen man vergessen hat. Ein Profil gab es
 * gar nicht: Ein Druck auf eine Zeile wählte, mehr war nicht.
 *
 * Die Anlage liefert die fünf Profilfelder seit dem 09.09.2026
 * (`pbx/src/client/server.ts`, Weg `/kollegen`; festgehalten in
 * `pbx/docs/CLIENT-API.md` und im Prüfstück `benutzer.profil.ts`).
 *
 * DAS GEGENSTÜCK AUF iOS HEISST GENAUSO und prüft dasselbe.
 *
 * WARUM QUELLTEXT-VERTRAG und kein Gerätetest: Ein Profilblatt lässt
 * sich ohne Emulator nicht öffnen, und der steht in dieser Umgebung
 * nicht zur Verfügung. Ein Vertrag, der den Fehler fängt, ist mehr wert
 * als ein Gerätetest, den niemand laufen lässt. Was er NICHT prüft:
 * ob es auf dem Gerät gut AUSSIEHT – das bleibt der Nachweis am Gerät.
 *
 * Aufruf:  node pruefstuecke/profil.mjs
 */
import { readFileSync, existsSync } from "node:fs";

const ergebnisse = [];
const pruefe = (name, ok, extra = "") => {
  ergebnisse.push([name, ok]);
  console.log(`${ok ? " ok  " : "FEHL "} ${name}${extra ? `  – ${extra}` : ""}`);
};
const lies = (p) => (existsSync(p) ? readFileSync(p, "utf8") : "");

const QUELLE = "app/src/main/java/de/dariatech/softphone";
const dienst = lies(`${QUELLE}/Dienst.kt`);
const profil = lies(`${QUELLE}/Profil.kt`);
const adapter = lies(`${QUELLE}/KontakteAdapter.kt`);
const haupt = lies(`${QUELLE}/MainActivity.kt`);
const texte = lies("app/src/main/res/values/strings.xml");

console.log("\n1) Die App liest die fünf Profilfelder");
/* GESUCHT WIRD IM BLOCK „Kollege", nicht in der ganzen Datei.

   Der erste Anlauf suchte in Dienst.kt insgesamt – und „ausserDienst"
   war grün, weil es die Klasse `Dienststand` schon gibt. Eine Prüfung,
   die den Treffer woanders findet, prüft nichts. */
const kollegeBlock =
  (dienst.match(/data class Kollege\([\s\S]*?\n\}/) ?? [""])[0];
pruefe("die Klasse Kollege ist auffindbar", kollegeBlock.length > 0,
  "ohne sie prüft alles Folgende ins Leere");
for (const feld of ["position", "abteilung", "email", "ausserDienst", "fax"]) {
  pruefe(`Kollege kennt „${feld}“`,
    new RegExp(`val\\s+${feld}\\s*:`).test(kollegeBlock),
    "sonst kommt es in der App nie an, auch wenn es im Portal steht");
}

/* UNBEKANNTE UND FEHLENDE FELDER DÜRFEN NICHT WEHTUN. Die Regel steht
   in CLIENT-API.md: „Unbekannte Felder werden ignoriert, nie als Fehler
   behandelt." Eine App im Feld lässt sich nicht mit dem Server
   gleichzeitig aktualisieren – eine ältere Anlage schickt diese Felder
   noch gar nicht. `optString`/`optBoolean` liefern dann leer bzw.
   false; `getString` würfe. */
const ausBlock = (dienst.match(/fun aus\(o: JSONObject\): Kollege \{[\s\S]*?\n {8}\}/) ?? [""])[0];
for (const feld of ["position", "abteilung", "email", "fax"]) {
  pruefe(`„${feld}“ wird mit optString gelesen`,
    new RegExp(`optString\\("${feld}"`).test(ausBlock),
    "eine ältere Anlage schickt es nicht – getString würfe");
}
pruefe("„ausserDienst“ wird mit optBoolean gelesen",
  /optBoolean\("ausserDienst"/.test(ausBlock), ausBlock ? "" : "kein aus()-Block gefunden");

console.log("\n2) Es gibt ein Profilblatt");
pruefe("die Datei Profil.kt ist da", profil.length > 0,
  "ohne sie gibt es kein Profil – nur eine Liste, die wählt");
pruefe("es zeigt das Profil eines KOLLEGEN",
  /fun\s+zeigeKollege|fun\s+zeige\(/.test(profil));
/* DAS EIGENE PROFIL, ausdrücklich verlangt: „soll im einegnen Profil
   als auch im Profil andere Koolegen". Es kommt aus Verzeichnis.ich() –
   die App weiß über ihre Nebenstelle, welcher Mensch sie ist. */
/* IM PROFILBLATT, nicht irgendwo in der App: MainActivity ruft
   Verzeichnis.ich() längst für das eigene Bild. Der erste Anlauf suchte
   in beiden Dateien – und war grün, bevor es ein Profil gab. */
pruefe("und das EIGENE, über Verzeichnis.ich",
  /Verzeichnis\.ich\(/.test(profil),
  "sonst sieht man jeden ausser sich selbst");

console.log("\n3) Im Profil steht, was im Portal steht");
for (const [feld, muster] of [
  ["Position", /\bposition\b/],
  ["Abteilung", /\babteilung\b/],
  ["E-Mail", /\bemail\b/],
  ["Faxnummer", /\bfax\b/],
  ["außer Dienst", /ausserDienst/],
  ["Durchwahl", /durchwahl/],
  ["Rufnummer", /\bnummer\b/]
]) {
  pruefe(`das Profil zeigt ${feld}`, muster.test(profil), "steht nicht in Profil.kt");
}

/* WAS NICHT INS PROFIL GEHÖRT. Die Anlage gibt es gar nicht erst
   heraus (benutzer.profil.ts hält das fest) – aber wer hier eines
   dieser Wörter findet, hat sich einen zweiten Weg gebaut. */
console.log("\n4) Und was NICHT hineingehört");
pruefe("keine Klingeldauer im Profil", !/klingelSec/.test(profil),
  "eine Einstellung der Technik, kein Profil");
pruefe("keine ausgehende Rufnummer im Profil", !/outboundNumber/.test(profil),
  "sie sagt, was der ANGERUFENE sieht – geht keinen Kollegen etwas an");
pruefe("kein Passwort", !/passwort|password/i.test(profil));

console.log("\n5) Die Liste zeigt, wonach man sucht");
pruefe("die Kollegenzeile nennt Position und Abteilung",
  /position/.test(adapter) && /abteilung/.test(adapter),
  "sonst muss man jede Zeile antippen, um zu sehen, wer das ist");
/* WER AUSSER DIENST IST, soll das in der Liste sehen – bevor er
   dreimal anruft. */
pruefe("und wer außer Dienst ist, steht als solcher da",
  /ausserDienst/.test(adapter));

console.log("\n6) Gesucht wird über die neuen Felder");
/* GENAU DAFÜR SIND SIE DA. Ein Feld, das die App anzeigt, aber nicht
   durchsucht, erfüllt den Auftrag zur Hälfte. */
const sucht = (profil + haupt + adapter).match(/fun\s+passt[\s\S]{0,600}/) ?? [""];
for (const feld of ["name", "position", "abteilung", "durchwahl"]) {
  pruefe(`die Suche findet über „${feld}“`,
    new RegExp(`\\b${feld}\\b`).test(sucht[0]),
    sucht[0] ? "" : "keine Suchfunktion gefunden");
}
pruefe("klein und groß geschrieben ist dasselbe",
  /lowercase\(\)/.test(sucht[0]), "wer „mta“ tippt, meint „MTA“");

console.log("\n7) Oberflächentexte stehen in strings.xml");
/* Dieselbe Regel wie überall in dieser App: Ein deutscher Satz, der im
   Kotlin steht, ist beim nächsten Mal an zwei Stellen verschieden. */
for (const name of ["profil_position", "profil_abteilung", "profil_ausser_dienst"]) {
  pruefe(`der Text „${name}“ ist da`, texte.includes(`"${name}"`));
}

const fehler = ergebnisse.filter(([, ok]) => !ok);
console.log(`\n${ergebnisse.length - fehler.length}/${ergebnisse.length} Prüfungen bestanden`);
process.exit(fehler.length > 0 ? 1 : 0);

File: app/src/main/java/de/dariatech/softphone/Profil.kt
package de.dariatech.softphone

import android.app.Activity
import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import com.google.android.material.bottomsheet.BottomSheetDialog
import com.google.android.material.button.MaterialButton

/**
 * Das Profil eines Menschen – das eigene und das der Kollegen.
 *
 * DER AUFTRAG (Inhaber, 09.09.2026): „jetzt alles was in Portal
 * eingetragen wird pro User soll in der App im Profil angezeogt werden,
 * soll im einegnen Profil als auch im Profil andere Koolegen der
 * Organistation." Als Vorbild hat er WhatsApp und Teams genannt.
 *
 * DER SCHADEN VORHER: Die Kollegenliste kannte vier Felder, und ein
 * Druck auf eine Zeile wählte. Mehr war nicht. Wer wissen wollte, wer
 * da eigentlich sitzt, musste jemanden fragen.
 *
 * ALLES KOMMT AUS `/kollegen`, nichts wird hier nachgeladen. Das
 * Verzeichnis liegt schon im Gerät (Verzeichnis.kt) – ein Profil, das
 * beim Öffnen erst den Server fragt, steht im Funkloch leer da.
 *
 * WAS HIER NICHT STEHT, und das ist Absicht: die Klingeldauer und die
 * ausgehende Rufnummer. Die Anlage gibt beides gar nicht erst heraus
 * (`pbx/src/client/server.ts`); es sind Einstellungen der Technik und
 * kein Profil. Wer sie sähe, hielte sie für etwas, das er ändern kann.
 *
 * DASSELBE GIBT ES AUF iOS (`Profil.swift`), mit denselben Feldern und
 * denselben Regeln.
 */
object Profil {

    /**
     * Das Profil eines Kollegen zeigen.
     *
     * `beimAnruf` und `beiNachricht` kommen von aussen, weil beides in
     * der MainActivity hängt (Anrufbereich, Gesprächsbildschirm). Ein
     * Blatt, das selbst anruft, müsste den halben Anrufweg kennen.
     */
    fun zeige(
        activity: Activity,
        wer: Kollege,
        ich: Boolean = false,
        beimAnruf: ((String) -> Unit)? = null,
        beiNachricht: ((Kollege) -> Unit)? = null
    ) {
        val blatt = LayoutInflater.from(activity).inflate(R.layout.blatt_profil, null)
        val dialog = BottomSheetDialog(activity)
        dialog.setContentView(blatt)

        val name = blatt.findViewById<TextView>(R.id.profilName)
        val rolle = blatt.findViewById<TextView>(R.id.profilRolle)
        val kuerzel = blatt.findViewById<TextView>(R.id.profilKuerzel)
        val foto = blatt.findViewById<ImageView>(R.id.profilFoto)
        val ausser = blatt.findViewById<TextView>(R.id.profilAusserDienst)
        val angaben = blatt.findViewById<LinearLayout>(R.id.profilAngaben)
        val leer = blatt.findViewById<TextView>(R.id.profilLeer)
        val knoepfe = blatt.findViewById<LinearLayout>(R.id.profilKnoepfe)
        val anrufen = blatt.findViewById<MaterialButton>(R.id.profilAnrufen)
        val nachricht = blatt.findViewById<MaterialButton>(R.id.profilNachricht)

        name.text = wer.name

        /* POSITION UND ABTEILUNG unter dem Namen, nicht in der Liste
           der Angaben: Sie sagen, WER das ist – dieselbe Stelle, an der
           bei WhatsApp der Statustext steht. */
        val rollentext = listOf(wer.position, wer.abteilung).filter { it.isNotEmpty() }
            .joinToString(" · ")
        rolle.text = rollentext
        rolle.visibility = if (rollentext.isEmpty()) View.GONE else View.VISIBLE

        /* BILD ODER INITIALEN – immer genau eines. Ein leerer Kreis
           sieht aus wie ein Bild, das nicht geladen hat. */
        val bild = Verzeichnis.bild(wer.id)
        if (bild != null) {
            foto.setImageBitmap(bild)
            foto.visibility = View.VISIBLE
            kuerzel.visibility = View.INVISIBLE
        } else {
            foto.visibility = View.GONE
            kuerzel.visibility = View.VISIBLE
            kuerzel.text = Verzeichnis.kuerzel(wer.name)
        }

        ausser.visibility = if (wer.ausserDienst) View.VISIBLE else View.GONE

        /* IM EIGENEN PROFIL KEINE KNÖPFE. Sich selbst anzurufen oder
           sich selbst zu schreiben ist kein Vorgang; die Anlage würde
           den Anruf durchstellen und das eigene Gerät klingeln lassen. */
        knoepfe.visibility = if (ich) View.GONE else View.VISIBLE

        val ziel = ruf(wer)
        anrufen.isEnabled = ziel.isNotEmpty()
        anrufen.setOnClickListener {
            dialog.dismiss()
            if (ziel.isNotEmpty()) beimAnruf?.invoke(ziel)
        }
        nachricht.setOnClickListener {
            dialog.dismiss()
            beiNachricht?.invoke(wer)
        }

        /* DIE ANGABEN. Nur, was gepflegt ist – eine Zeile „Abteilung: –"
           sagt nichts und kostet Platz. */
        angaben.removeAllViews()
        val ctx = activity as Context
        zeile(ctx, angaben, R.string.profil_position, wer.position)
        zeile(ctx, angaben, R.string.profil_abteilung, wer.abteilung)
        zeile(ctx, angaben, R.string.profil_durchwahl, wer.durchwahl)
        zeile(ctx, angaben, R.string.profil_nummer, wer.nummer)
        zeile(ctx, angaben, R.string.profil_fax, wer.fax)
        zeile(ctx, angaben, R.string.profil_email, wer.email)

        /* IST NICHTS GEPFLEGT, sagt das Blatt das – statt eine leere
           Fläche zu zeigen. „Nichts da" und „kaputt" sehen sonst gleich
           aus. Im eigenen Profil steht dazu, WER es einträgt. */
        if (angaben.childCount == 0) {
            leer.visibility = View.VISIBLE
            leer.setText(if (ich) R.string.profil_ich_leer else R.string.profil_leer)
        } else {
            leer.visibility = View.GONE
        }

        dialog.show()
    }

    /**
     * Das EIGENE Profil.
     *
     * Wer man ist, weiß die App über ihre Nebenstelle (Verzeichnis.ich).
     * Gehört dieser Apparat keinem Menschen – ein Konferenzraum, ein
     * Türsprecher –, ist das keine Störung, sondern eine Auskunft.
     */
    fun zeigeMich(activity: Activity) {
        val ich = Verzeichnis.ich(activity)
        if (ich == null) {
            android.app.AlertDialog.Builder(activity)
                .setTitle(R.string.profil_ich)
                .setMessage(R.string.profil_kein_ich)
                .setPositiveButton(android.R.string.ok, null)
                .show()
            return
        }
        zeige(activity, ich, ich = true)
    }

    /**
     * Wen ruft man bei einem Kollegen an? SEINE DURCHWAHL, wenn er eine
     * hat – dann klingeln ALLE seine Geräte. Sonst die erste
     * Nebenstelle, also ein einzelnes Gerät. Dieselbe Regel wie in der
     * Kontaktliste (KontakteAdapter); zwei Fassungen davon laufen
     * auseinander.
     */
    fun ruf(k: Kollege): String = when {
        k.durchwahl.isNotEmpty() -> k.durchwahl
        k.nebenstellen.isNotEmpty() -> k.nebenstellen.first()
        else -> k.nummer
    }

    /**
     * Passt dieser Mensch zur Suche?
     *
     * GENAU DAFÜR SIND POSITION UND ABTEILUNG DA – der Auftrag lautete
     * „um in Apps nach der mitarbeiter zu suchen". Ein Feld, das die App
     * anzeigt, aber nicht durchsucht, erfüllt ihn zur Hälfte.
     *
     * Klein geschrieben verglichen: Wer „mta" tippt, meint „MTA".
     */
    fun passt(k: Kollege, suche: String): Boolean {
        val q = suche.trim().lowercase()
        if (q.isEmpty()) return true
        return listOf(k.name, k.position, k.abteilung, k.durchwahl, k.nummer, k.email)
            .any { it.lowercase().contains(q) }
    }

    /** Eine Zeile „Beschriftung / Wert" – oder gar keine, wenn leer. */
    private fun zeile(ctx: Context, wohin: LinearLayout, beschriftung: Int, wert: String) {
        if (wert.isEmpty()) return
        val spalte = LinearLayout(ctx).apply {
            orientation = LinearLayout.VERTICAL
            layoutParams = LinearLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.WRAP_CONTENT
            ).apply { topMargin = 14 }
        }
        spalte.addView(TextView(ctx).apply {
            setText(beschriftung)
            textSize = 12f
            setTextColor(farbe(ctx, com.google.android.material.R.attr.colorOnSurfaceVariant))
        })
        spalte.addView(TextView(ctx).apply {
            text = wert
            textSize = 16f
            setTextIsSelectable(true)
            setTextColor(farbe(ctx, com.google.android.material.R.attr.colorOnSurface))
        })
        wohin.addView(spalte)
    }

    private fun farbe(ctx: Context, attr: Int): Int {
        val wert = android.util.TypedValue()
        ctx.theme.resolveAttribute(attr, wert, true)
        return wert.data
    }
}
